LinuxAdapter.cs

```diff
@@ -0,0 +1,120 @@
+            _connection.Bind(); // This line actually makes the connection.
+            _connection.Timeout = TimeSpan.FromMinutes(1);
+        }
+
+        // example: machineAccount = "user@DOMAIN.com" machinePassword = "***"
+        public Task OnAuthenticated(AuthenticatedContext context, bool resolveNestedGroups = true)
+        {
+            var loggerFactory = context.HttpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger("LinuxAdapter");
+
+            var user = context.Principal.Identity.Name;
+            var userAccountName = user.Substring(0, user.IndexOf('@'));
+
+            string filter = $"(&(objectClass=user)(sAMAccountName={userAccountName}))"; // This is using ldap search query language, it is looking on the server for someUser
+            SearchRequest searchRequest = new SearchRequest(_distinguishedName, filter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)_connection.SendRequest(searchRequest);
+
+            if (searchResponse.Entries.Count > 0)
+            {
+                if (searchResponse.Entries.Count > 1)
+                {
+                    logger.LogWarning($"More than one response received for query: {filter} with distinguished name: {_distinguishedName}");
+                }
+
+                var userFound = searchResponse.Entries[0]; //Get the object that was found on ldap
+                var memberof = userFound.Attributes["memberof"]; // You can access ldap Attributes with Attributes property
+
+                var claimsIdentity = context.Principal.Identity as ClaimsIdentity;
+
+                foreach (var group in memberof)
+                {
+                    // Example distinguished name: CN=TestGroup,DC=KERB,DC=local
+                    var groupDN = $"{Encoding.UTF8.GetString((byte[])group)}";
+                    var groupCN = groupDN.Split(',')[0].Substring("CN=".Length);
+
+                    if (resolveNestedGroups)
+                    {
+                        GetNestedGroups(claimsIdentity, groupCN, logger);
+                    }
+                    else
+                    {
+                        AddRole(claimsIdentity, groupCN);
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void GetNestedGroups(ClaimsIdentity principal, string groupCN, ILogger logger)
+        {
+            string filter = $"(&(objectClass=group)(sAMAccountName={groupCN}))"; // This is using ldap search query language, it is looking on the server for someUser
+            SearchRequest searchRequest = new SearchRequest(_distinguishedName, filter, System.DirectoryServices.Protocols.SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)_connection.SendRequest(searchRequest);
+
+            if (searchResponse.Entries.Count > 0)
```

Comment:
Should log the 0 case too. Wouldn't that mean you don't add that group as a role at all?

Reply:
I'll add a log.

Reply:
If the group doesn't exist, should it still be added to roles?

Reply:
@Tratcher @joperezr?

Reply:
I'd say no, it doesn't get added.